### PR TITLE
[Typechecker] Allow @objc functions to return dynamic self

### PR DESCRIPTION
**Pull request by theblixguy**
This PR lifts a restriction on @objc functions which disallowed dynamic self as return type.

```swift
@objc protocol Foo {
    static func makeFoo() -> Self // ok
}
```
Resolves [SR-7601](https://bugs.swift.org/browse/SR-7601).

### CODE REVIEW

**Comment by theblixguy** [comment]
cc @jrose-apple

**Comment by jrose-apple** [comment]
This seems correct to me, but it's worth adding a PrintAsObjC test too to make sure it gets printed as `instancetype`. @slavapestov, any additional comments?

File: test/PrintAsObjC/dynamicself.swift [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+  func fooFunc() -> Self
+}
+
+// CHECK-LABEL: @interface BarClass : NSObject <FooProtocol>
+// CHECK-NEXT:    - (nonnull instancetype)fooFunc SWIFT_WARN_UNUSED_RESULT;
+// CHECK-NEXT:    - (nonnull instancetype)init OBJC_DESIGNATED_INITIALIZER;
+// CHECK-NEXT:  @end
+
+@objc @objcMembers class BarClass: NSObject, FooProtocol {
+  func fooFunc() -> Self { return self }
```
Comment (jrose-apple) [inline test/PrintAsObjC/dynamicself.swift]: Can you add one that returns `Optional<Self>` too? Just to be thorough!

Comment (theblixguy) [reply test/PrintAsObjC/dynamicself.swift]: @jrose-apple done 👍

**Review by jrose-apple — APPROVED** [review]

**Comment by theblixguy** [comment]
Seems like there's a crash: `If you're seeing a crash here, check that your SDK and dependencies are at least as new as the versions used to build 'dynamicself'`

@jrose-apple do you know what could be the issue here?

**Comment by jrose-apple** [comment]
Ah, sorry. `%clang-importer-sdk` doesn't work with anything past SILGen; you'll have to copy over the "-enable-source-import hackaround" lines from one of the other PrintAsObjC tests. Or just move this into existing tests (the most boring ones being classes.swift and protocols.swift).

Sorry for not catching that in the review.

**Comment by theblixguy** [comment]
@jrose-apple oh, it's alright! I have updated the test file. Can you take a look and invoke the CI again if it looks good to you?

File: test/PrintAsObjC/dynamicself.swift [hunk 2]
```diff
@@ -1,4 +1,11 @@
 // RUN: %empty-directory(%t)
+
+// FIXME: BEGIN -enable-source-import hackaround
+// RUN:  %target-swift-frontend(mock-sdk: -sdk %S/../Inputs/clang-importer-sdk -I %t) -emit-module -o %t  %S/../Inputs/clang-importer-sdk/swift-modules/ObjectiveC.swift -disable-objc-attr-requires-foundation-module
+// RUN:  %target-swift-frontend(mock-sdk: -sdk %S/../Inputs/clang-importer-sdk -I %t) -emit-module -o %t  %S/../Inputs/clang-importer-sdk/swift-modules/CoreGraphics.swift
+// RUN:  %target-swift-frontend(mock-sdk: -sdk %S/../Inputs/clang-importer-sdk -I %t) -emit-module -o %t  %S/../Inputs/clang-importer-sdk/swift-modules/Foundation.swift
+// FIXME: END -enable-source-import hackaround
+
 // RUN: %target-swift-frontend(mock-sdk: %clang-importer-sdk) -emit-module -o %t %s -module-name dynamicself -disable-objc-attr-requires-foundation-module
 // RUN: %target-swift-frontend(mock-sdk: %clang-importer-sdk) -parse-as-library %t/dynamicself.swiftmodule -typecheck -emit-objc-header-path %t/dynamicself.h -disable-objc-attr-requires-foundation-module
```
[the remark this answers is not in the document]
Comment (theblixguy) [reply test/PrintAsObjC/dynamicself.swift]: @jrose-apple Done. I ran the test locally and it passes 🎉 can you invoke the CI? thanks!

**Comment by jrose-apple** [comment]
@swift-ci Please test

**Comment by theblixguy** [comment]
All tests have passed 🎉 @jrose-apple

**Comment by jrose-apple** [comment]
Thank you for the implementation!

**Pull request merged**